File: src/paniolo/_netif.py
# Copyright 2026 Curtis Galloway
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

"""Switch the target's USB-Ethernet link between netboot and ffx modes.

Both modes share one physical point-to-point link (the USB-Ethernet dongle to
the Pi's RP1 GEM). They are mutually exclusive:

  netboot — IPv4 ``host_ip``/24 + DHCP + TFTP (the daemons managed by
            ``_netboot``). The Pi TFTP-boots a kernel + ramdisk.

  ffx     — IPv6 link-local (``fe80::1``/64) on the host interface, with **no**
            DHCP/TFTP. The Pi boots from its SD card and is reached over ffx at
            ``fe80::<dev-slaac>%<iface>``.

``netif mode`` makes the transition atomic and removes the two highest-cost
seams of doing it by hand:

  1. Switching to ``ffx`` stops netboot *first*, so the next power-cycle falls
     through to the SD card instead of silently TFTP-booting a stale image.
  2. Switching to ``ffx`` adds the host-side ``fe80::1``/64 that ffx needs and
     that nothing else sets up — no manual ``sudo ip -6 addr add``.

Every mode is idempotent and re-runnable. The IPv6 link-local is ephemeral
(lost on a control-host reboot), so ``mode ffx`` simply re-adds it when absent.
The active mode is *probed* (from the running daemons and the interface's
addresses), not stored, so it stays correct even after a reboot clears things.
"""

from __future__ import annotations

import subprocess
import sys

from ._config import TargetConfig
from . import _netboot
from ._state import is_netboot_running

# The host-side IPv6 link-local that ffx talks through. Any LL works on a
# point-to-point link; ::1 is the conventional, easy-to-type choice and matches
# the manual recipe (`sudo ip -6 addr add fe80::1/64 dev <iface>`).
FFX_HOST_LL = "fe80::1"
FFX_PREFIX = 64

MODES = ("netboot", "ffx", "off")

_SUDO_HINT = (
    "Ensure passwordless sudo is configured (NOPASSWD) for the control machine."
)


# ── interface address probing ──────────────────────────────────────────────


def iface_addresses(interface: str) -> dict:
    """Return the interface's assigned addresses as {"inet": [...], "inet6": [...]}.

    IPv6 addresses keep their scope suffix stripped (``fe80::1%enx0`` → ``fe80::1``)
    but retain any prefix length (``fe80::1/64``) so callers can match exactly.
    """
    inet: list[str] = []
    inet6: list[str] = []
    if sys.platform == "darwin":
        result = subprocess.run(["ifconfig", interface], capture_output=True, text=True)
        for line in result.stdout.splitlines():
            s = line.strip()
            if s.startswith("inet "):
                inet.append(s.split()[1])
            elif s.startswith("inet6 "):
                inet6.append(s.split()[1].split("%", 1)[0])
        return {"inet": inet, "inet6": inet6}

    result = subprocess.run(
        ["ip", "-brief", "addr", "show", "dev", interface],
        capture_output=True,
        text=True,
    )
    if result.returncode == 0:
        # `ip -brief` columns: <iface> <state> <addr> <addr> ...
        for addr in result.stdout.split()[2:]:
            (inet6 if ":" in addr else inet).append(addr)
    return {"inet": inet, "inet6": inet6}


def _has_host_ll(addrs: dict) -> bool:
    """True if the ffx host link-local is currently assigned to the interface."""
    return any(a.split("/", 1)[0] == FFX_HOST_LL for a in addrs["inet6"])


def ipv6_peers(interface: str) -> list[str]:
    """Link-local IPv6 neighbours discovered on the interface (Linux only).

    Surfaces the device's own ``fe80::…`` address for a ready-to-paste
    ``ffx target add fe80::…%<iface>`` without scraping the serial log. Excludes
    our own host LL. Returns [] on macOS or when nothing has been discovered.
    """
    if sys.platform == "darwin":
        return []
    result = subprocess.run(
        ["ip", "-6", "neigh", "show", "dev", interface],
        capture_output=True,
        text=True,
    )
    if result.returncode != 0:
        return []
    peers: list[str] = []
    for line in result.stdout.splitlines():
        toks = line.split()
        if toks and toks[0].startswith("fe80:") and toks[0] != FFX_HOST_LL:
            peers.append(toks[0])
    return peers


# ── privileged interface mutation ──────────────────────────────────────────


def _add_host_ll(interface: str) -> None:
    """Add the ffx host link-local to the interface (idempotent)."""
    if sys.platform == "darwin":
        # macOS auto-assigns a link-local already; aliasing in fe80::1 is
        # best-effort and harmless if it already exists.
        subprocess.run(
            [
                "sudo",
                "ifconfig",
                interface,
                "inet6",
                FFX_HOST_LL,
                "prefixlen",
                str(FFX_PREFIX),
                "up",
            ],
            capture_output=True,
            text=True,
            check=False,
        )
        return

    # IPv6 may be disabled on a fresh secondary interface; enable it. Use the
    # '/'-separated sysctl key form so interface names are passed literally.
    subprocess.run(
        ["sudo", "sysctl", "-w", f"net/ipv6/conf/{interface}/disable_ipv6=0"],
        capture_output=True,
        text=True,
        check=False,
    )
    subprocess.run(["sudo", "ip", "link", "set", interface, "up"], check=False)
    result = subprocess.run(
        [
            "sudo",
            "ip",
            "-6",
            "addr",
            "add",
            f"{FFX_HOST_LL}/{FFX_PREFIX}",
            "dev",
            interface,
        ],
        capture_output=True,
        text=True,
    )
    if result.returncode != 0 and "exists" not in result.stderr.lower():
        raise RuntimeError(
            f"ip -6 addr add {FFX_HOST_LL}/{FFX_PREFIX} dev {interface} failed: "
            f"{result.stderr.strip()}\n{_SUDO_HINT}"
        )


def _del_host_ll(interface: str) -> None:
    """Remove the ffx host link-local from the interface (best-effort)."""
    if sys.platform == "darwin":
        subprocess.run(
            ["sudo", "ifconfig", interface, "inet6", FFX_HOST_LL, "-alias"],
            capture_output=True,
            text=True,
            check=False,
        )
        return
    subprocess.run(
        [
            "sudo",
            "ip",
            "-6",
            "addr",
            "del",
            f"{FFX_HOST_LL}/{FFX_PREFIX}",
            "dev",
            interface,
        ],
        capture_output=True,
        text=True,
        check=False,
    )


def _del_host_ip(interface: str, host_ip: str) -> None:
    """Remove a lingering netboot IPv4 from the interface (Linux, best-effort).

    Targets only ``host_ip``/24 so we never touch an unrelated address. macOS
    teardown goes through ``networksetup -setdhcp`` in ``_netboot`` instead.
    """
    if sys.platform == "darwin":
        return
    subprocess.run(
        ["sudo", "ip", "addr", "del", f"{host_ip}/24", "dev", interface],
        capture_output=True,
        text=True,
        check=False,
    )


# ── mode transitions ───────────────────────────────────────────────────────


def mode_netboot(cfg: TargetConfig, engine: str = "rust") -> None:
    """Put the link in netboot mode: tear down ffx, start DHCP+TFTP.

    Idempotent: if netboot is already running it is left as-is (the ffx
    link-local is still cleared). ``_netboot.start`` itself enforces the
    primary-NIC guard and configures the IPv4 host address.
    """
    _del_host_ll(cfg.interface)
    if is_netboot_running(cfg.name):
        return
    _netboot.start(cfg, engine=engine)


def mode_ffx(cfg: TargetConfig) -> None:
    """Put the link in ffx mode: stop netboot, add the host IPv6 link-local.

    Stopping netboot first is the point of this command — it means the next
    power-cycle boots from SD instead of TFTP. Re-running re-adds the ephemeral
    link-local if a control-host reboot cleared it.
    """
    if is_netboot_running(cfg.name):
        _netboot.stop(cfg.name)
    _add_host_ll(cfg.interface)


def mode_off(cfg: TargetConfig) -> None:
    """Drop both configs: stop netboot, remove the ffx LL and any stale IPv4."""
    if is_netboot_running(cfg.name):
        # _netboot.stop restores the interface (flushes the IPv4 host address).
        _netboot.stop(cfg.name)
    _del_host_ll(cfg.interface)
    _del_host_ip(cfg.interface, cfg.host_ip)


# ── status ─────────────────────────────────────────────────────────────────


def get_status(cfg: TargetConfig) -> dict:
    """Probe and report the active mode plus the interface's addresses.

    ``mode`` is derived, not stored: netboot daemons running → ``netboot``; else
    the ffx host LL present → ``ffx``; else ``off``.
    """
    netboot_running = is_netboot_running(cfg.name)
    addrs = iface_addresses(cfg.interface)
    has_ll = _has_host_ll(addrs)

    if netboot_running:
        mode = "netboot"
    elif has_ll:
        mode = "ffx"
    else:
        mode = "off"

    return {
        "target": cfg.name,
        "interface": cfg.interface,
        "mode": mode,
        "netboot_running": netboot_running,
        "host_ip": cfg.host_ip,
        "host_ll": FFX_HOST_LL if has_ll else None,
        "inet": addrs["inet"],
        "inet6": addrs["inet6"],
        "peers": ipv6_peers(cfg.interface) if mode == "ffx" else [],
    }

File: cli/src/netif.rs
// Copyright 2026 Curtis Galloway
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

//! USB-Ethernet link plumbing for the netboot/ffx modes.
//!
//! Ported from `_netboot.py` (interface helpers) + `_netif.py` (modes). The two
//! link modes share one physical point-to-point link and are mutually
//! exclusive: **netboot** = IPv4 host_ip/24 + DHCP/TFTP; **ffx** = host IPv6
//! link-local `fe80::1/64`, no daemons. The active mode is probed from the
//! running daemon + the interface's addresses, never stored.

use std::process::Command;

use anyhow::{bail, Result};

/// The host-side IPv6 link-local that ffx talks through.
pub const FFX_HOST_LL: &str = "fe80::1";
pub const FFX_PREFIX: u8 = 64;

const SUDO_HINT: &str =
    "Ensure passwordless sudo is configured (NOPASSWD) for the control machine.";

fn run(cmd: &[&str]) -> std::io::Result<std::process::Output> {
    Command::new(cmd[0]).args(&cmd[1..]).output()
}

fn macos() -> bool {
    cfg!(target_os = "macos")
}

// ── interface discovery / probing ───────────────────────────────────────────

pub struct EthInterface {
    pub port: String,
    pub device: String,
    pub active: bool,
}

pub fn is_interface_active(device: &str) -> bool {
    if macos() {
        run(&["ifconfig", device])
            .map(|o| String::from_utf8_lossy(&o.stdout).contains("status: active"))
            .unwrap_or(false)
    } else {
        std::fs::read_to_string(format!("/sys/class/net/{device}/carrier"))
            .map(|s| s.trim() == "1")
            .unwrap_or(false)
    }
}

const MAC_EXCLUDE_PORTS: [&str; 6] = [
    "Wi-Fi",
    "Thunderbolt",
    "Bluetooth",
    "FireWire",
    "iPhone",
    "iPad",
];
const LINUX_SKIP_PREFIXES: [&str; 8] = [
    "lo", "docker", "veth", "br", "virbr", "vlan", "bond", "dummy",
];

/// External (non-built-in) Ethernet interfaces, active ones first. The
/// interface carrying the system default route is excluded — it can never be a
/// netboot link (the start guard refuses to reconfigure the primary NIC), so
/// surfacing it in discovery only invites a misconfiguration.
pub fn list_usb_ethernet_interfaces() -> Vec<EthInterface> {
    let primary = default_route_interface();
    let mut out: Vec<EthInterface> = Vec::new();
    if macos() {
        let Ok(o) = run(&["networksetup", "-listallhardwareports"]) else {
            return out;
        };
        let text = String::from_utf8_lossy(&o.stdout).into_owned();
        let mut port: Option<String> = None;
        for line in text.lines() {
            if let Some(p) = line.strip_prefix("Hardware Port:") {
                port = Some(p.trim().to_string());
            } else if let Some(d) = line.strip_prefix("Device:") {
                if let Some(p) = port.take() {
                    let device = d.trim().to_string();
                    let excluded = device == "bridge0"
                        || device == "lo0"
                        || primary.as_deref() == Some(device.as_str())
                        || MAC_EXCLUDE_PORTS.iter().any(|x| p.starts_with(x));
                    if !excluded {
                        let active = is_interface_active(&device);
                        out.push(EthInterface {
                            port: p,
                            device,
                            active,
                        });
                    }
                }
            }
        }
    } else if let Ok(rd) = std::fs::read_dir("/sys/class/net") {
        for e in rd.flatten() {
            let name = e.file_name().to_string_lossy().into_owned();
            if LINUX_SKIP_PREFIXES.iter().any(|p| name.starts_with(p))
                || primary.as_deref() == Some(name.as_str())
            {
                continue;
            }
            // Type 1 = Ethernet (ARPHRD_ETHER).
            let is_ether = std::fs::read_to_string(e.path().join("type"))
                .map(|t| t.trim() == "1")
                .unwrap_or(false);
            if !is_ether {
                continue;
            }
            // Wi-Fi also reports ARPHRD_ETHER, but a wireless NIC can never be
            // the point-to-point netboot link; the `wireless` sysfs dir marks it.
            if e.path().join("wireless").exists() {
                continue;
            }
            let active = is_interface_active(&name);
            out.push(EthInterface {
                port: name.clone(),
                device: name,
                active,
            });
        }
    }
    out.sort_by(|a, b| (!a.active, &a.device).cmp(&(!b.active, &b.device)));
    out
}

/// The networksetup service name for a device (macOS only).
fn find_network_service(interface: &str) -> Option<String> {
    if !macos() {
        return None;
    }
    let o = run(&["networksetup", "-listallhardwareports"]).ok()?;
    let text = String::from_utf8_lossy(&o.stdout).into_owned();
    let mut service: Option<String> = None;
    for line in text.lines() {
        if let Some(p) = line.strip_prefix("Hardware Port:") {
            service = Some(p.trim().to_string());
        } else if let Some(d) = line.strip_prefix("Device:") {
            if d.trim() == interface {
                return service;
            }
        }
    }
    None
}

/// The interface carrying the system default route, or None.
fn default_route_interface() -> Option<String> {
    if macos() {
        let o = run(&["route", "-n", "get", "default"]).ok()?;
        let text = String::from_utf8_lossy(&o.stdout).into_owned();
        for line in text.lines() {
            if let Some(i) = line.trim().strip_prefix("interface:") {
                return Some(i.trim().to_string());
            }
        }
        None
    } else {
        let o = run(&["ip", "route", "show", "default"]).ok()?;
        let text = String::from_utf8_lossy(&o.stdout).into_owned();
        let toks: Vec<&str> = text.split_whitespace().collect();
        toks.iter()
            .position(|&t| t == "dev")
            .and_then(|i| toks.get(i + 1))
            .map(|s| s.to_string())
    }
}

/// The pure decision behind [`is_primary_interface`]: `default_route` is the
/// interface carrying the system default route (if any), and `interface` is
/// primary when it is that one. Free of probing so it can be unit-tested.
fn is_default_route_interface(interface: &str, default_route: Option<&str>) -> bool {
    default_route == Some(interface)
}

/// True if `interface` carries the system default route — netboot and every
/// netif mode change must never reconfigure the primary NIC.
pub fn is_primary_interface(interface: &str) -> bool {
    is_default_route_interface(interface, default_route_interface().as_deref())
}

/// The refusal every mode change shares with `netboot::start`: never touch the
/// interface carrying the system default route. A lab file naming the host's
/// real NIC would otherwise let `mode link` replace the control host's only
/// address, or `down-hard` admin-down it — an SSH lockout on a remote control
/// host. Pure: `default_route` is what [`default_route_interface`] returned, so
/// both the decision and the message are unit-testable; `what` names the verb.
fn refuse_primary_interface(
    what: &str,
    interface: &str,
    default_route: Option<&str>,
) -> Result<()> {
    if is_default_route_interface(interface, default_route) {
        bail!(
            "refusing netif {what} on '{interface}': it carries the system default \
             route (your primary network interface). netif would reconfigure it and \
             break host networking. Use a dedicated USB-Ethernet adapter for the \
             netboot link."
        );
    }
    Ok(())
}

/// [`refuse_primary_interface`] against the live routing table.
fn guard_primary_interface(what: &str, interface: &str) -> Result<()> {
    refuse_primary_interface(what, interface, default_route_interface().as_deref())
}

/// The interface's addresses: (inet, inet6) with IPv6 scope suffixes stripped.
pub fn iface_addresses(interface: &str) -> (Vec<String>, Vec<String>) {
    let mut inet = Vec::new();
    let mut inet6 = Vec::new();
    if macos() {
        if let Ok(o) = run(&["ifconfig", interface]) {
            for line in String::from_utf8_lossy(&o.stdout).lines() {
                let s = line.trim();
                if let Some(rest) = s.strip_prefix("inet ") {
                    if let Some(a) = rest.split_whitespace().next() {
                        inet.push(a.to_string());
                    }
                } else if let Some(rest) = s.strip_prefix("inet6 ") {
                    if let Some(a) = rest.split_whitespace().next() {
                        inet6.push(a.split('%').next().unwrap_or(a).to_string());
                    }
                }
            }
        }
    } else if let Ok(o) = run(&["ip", "-brief", "addr", "show", "dev", interface]) {
        if o.status.success() {
            let text = String::from_utf8_lossy(&o.stdout).into_owned();
            for addr in text.split_whitespace().skip(2) {
                if addr.contains(':') {
                    inet6.push(addr.to_string());
                } else {
                    inet.push(addr.to_string());
                }
            }
        }
    }
    (inet, inet6)
}

fn has_host_ll(inet6: &[String]) -> bool {
    inet6
        .iter()
        .any(|a| a.split('/').next().unwrap_or(a) == FFX_HOST_LL)
}

/// Link-local IPv6 neighbours on the interface (Linux only; excludes our LL).
pub fn ipv6_peers(interface: &str) -> Vec<String> {
    if macos() {
        return Vec::new();
    }
    let Ok(o) = run(&["ip", "-6", "neigh", "show", "dev", interface]) else {
        return Vec::new();
    };
    if !o.status.success() {
        return Vec::new();
    }
    String::from_utf8_lossy(&o.stdout)
        .lines()
        .filter_map(|l| l.split_whitespace().next())
        .filter(|t| t.starts_with("fe80:") && *t != FFX_HOST_LL)
        .map(String::from)
        .collect()
}

// ── privileged interface mutation ───────────────────────────────────────────

/// What a failed command said: stderr, or stdout when stderr is empty
/// (`networksetup` reports its `** Error:` lines on stdout).
fn failure_text(o: &std::process::Output) -> String {
    let err = String::from_utf8_lossy(&o.stderr).trim().to_string();
    if err.is_empty() {
        String::from_utf8_lossy(&o.stdout).trim().to_string()
    } else {
        err
    }
}

/// The IPv4 addresses (`addr/prefix`, as printed) in `ip -4 -o addr show dev
/// <iface>` output: one address per line, the token after `inet`. Pure, so the
/// format is pinned by a test.
fn ipv4_addrs(show_output: &str) -> Vec<String> {
    show_output
        .lines()
        .filter_map(|line| {
            let mut toks = line.split_whitespace();
            toks.find(|t| *t == "inet")?;
            toks.next().map(str::to_string)
        })
        .collect()
}

/// Linux: `ip addr replace` — idempotent, and it lands the address before
/// anything else is touched.
fn ip_addr_replace(interface: &str, cidr: &str) -> Result<()> {
    let o = run(&["sudo", "ip", "addr", "replace", cidr, "dev", interface])?;
    if !o.status.success() {
        bail!(
            "ip addr replace {cidr} dev {interface} failed: {}\n{SUDO_HINT}",
            String::from_utf8_lossy(&o.stderr).trim()
        );
    }
    Ok(())
}

/// Assign the static netboot host IP to the interface (sudo).
pub fn configure_interface(interface: &str, host_ip: &str) -> Result<()> {
    if macos() {
        if let Some(service) = find_network_service(interface) {
            // Tell configd the service is manual, so a carrier flap (every
            // target power-cycle) doesn't hand the link back to DHCP.
            // `-setmanual` takes four arguments — service, ip, subnet, router —
            // and rejects the call without the router; the host IP itself is
            // the customary router for a point-to-point link with no gateway.
            // `ifconfig` below is what makes the address live, so a failure
            // here is a warning, not a failure of the call.
            let manual = run(&[
                "sudo",
                "networksetup",
                "-setmanual",
                &service,
                host_ip,
                "255.255.255.0",
                host_ip,
            ]);
            match manual {
                Ok(o) if o.status.success() => {}
                Ok(o) => eprintln!(
                    "warning: networksetup -setmanual '{service}' failed: {}",
                    failure_text(&o)
                ),
                Err(e) => eprintln!("warning: networksetup -setmanual '{service}' failed: {e}"),
            }
        }
        let o = run(&[
            "sudo",
            "ifconfig",
            interface,
            host_ip,
            "netmask",
            "255.255.255.0",
            "up",
        ])?;
        if !o.status.success() {
            bail!(
                "ifconfig {interface} failed: {}\n{SUDO_HINT}",
                String::from_utf8_lossy(&o.stderr).trim()
            );
        }
    } else {
        // Replace first, remove the rest after: a failed assignment leaves the
        // interface as it was, never addressless (flush-then-add did).
        let cidr = format!("{host_ip}/24");
        ip_addr_replace(interface, &cidr)?;
        let others: Vec<String> = match run(&["ip", "-4", "-o", "addr", "show", "dev", interface]) {
            Ok(o) => ipv4_addrs(&String::from_utf8_lossy(&o.stdout))
                .into_iter()
                .filter(|a| *a != cidr)
                .collect(),
            Err(_) => Vec::new(),
        };
        for other in &others {
            let _ = run(&["sudo", "ip", "addr", "del", other, "dev", interface]);
        }
        if !others.is_empty() {
            // Deleting a primary address takes its secondaries with it unless
            // promote_secondaries is on; re-assert ours either way.
            ip_addr_replace(interface, &cidr)?;
        }
        let _ = run(&["sudo", "ip", "link", "set", interface, "up"]);
    }
    Ok(())
}

/// Release the static IP and return the interface to OS-managed networking.
pub fn restore_interface(interface: &str) {
    if macos() {
        if let Some(service) = find_network_service(interface) {
            let _ = run(&["sudo", "networksetup", "-setdhcp", &service]);
        }
    } else {
        let _ = run(&["sudo", "ip", "addr", "flush", "dev", interface]);
    }
}

/// macOS: stop neighbour-unreachability detection from blocking sends to the
/// Pi's bootloader (it never answers ARP probes). No-op on Linux.
pub fn tune_arp_for_silent_client() {
    if !macos() {
        return;
    }
    for kv in [
        "net.link.ether.inet.arp_llreach_base=0",
        "net.link.ether.inet.host_down_time=0",
    ] {
        let _ = run(&["sudo", "sysctl", "-w", kv]);
    }
}

fn add_host_ll(interface: &str) -> Result<()> {
    if macos() {
        let o = run(&[
            "sudo",
            "ifconfig",
            interface,
            "inet6",
            FFX_HOST_LL,
            "prefixlen",
            "64",
            "up",
        ])?;
        let stderr = String::from_utf8_lossy(&o.stderr).into_owned();
        // Re-adding an address that is already there is the idempotent re-run.
        if !o.status.success() && !stderr.to_lowercase().contains("exists") {
            bail!(
                "ifconfig {interface} inet6 {FFX_HOST_LL} failed: {}\n{SUDO_HINT}",
                stderr.trim()
            );
        }
        return Ok(());
    }
    let sysctl_key = format!("net/ipv6/conf/{interface}/disable_ipv6=0");
    let _ = run(&["sudo", "sysctl", "-w", &sysctl_key]);
    let _ = run(&["sudo", "ip", "link", "set", interface, "up"]);
    let cidr = format!("{FFX_HOST_LL}/{FFX_PREFIX}");
    let o = run(&["sudo", "ip", "-6", "addr", "add", &cidr, "dev", interface])?;
    let stderr = String::from_utf8_lossy(&o.stderr).into_owned();
    if !o.status.success() && !stderr.to_lowercase().contains("exists") {
        bail!(
            "ip -6 addr add {cidr} dev {interface} failed: {}\n{SUDO_HINT}",
            stderr.trim()
        );
    }
    Ok(())
}

fn del_host_ll(interface: &str) {
    if macos() {
        let _ = run(&[
            "sudo",
            "ifconfig",
            interface,
            "inet6",
            FFX_HOST_LL,
            "-alias",
        ]);
    } else {
        let cidr = format!("{FFX_HOST_LL}/{FFX_PREFIX}");
        let _ = run(&["sudo", "ip", "-6", "addr", "del", &cidr, "dev", interface]);
    }
}

fn del_host_ip(interface: &str, host_ip: &str) {
    if macos() {
        // Release the static IP by returning the service to DHCP — an
        // `ifconfig` delete won't unset a `networksetup -setmanual` IP, so
        // `mode off` from `link` mode (no netboot stop to restore it) would
        // otherwise leave the host IP assigned.
        if let Some(service) = find_network_service(interface) {
            let _ = run(&["sudo", "networksetup", "-setdhcp", &service]);
        }
        // `-setdhcp` reconfigures the service, but the address `ifconfig` put
        // on the interface can outlive it as an alias; release it explicitly so
        // a stale host IP cannot survive `mode off`.
        let _ = run(&["sudo", "ifconfig", interface, "inet", host_ip, "-alias"]);
        return;
    }
    let cidr = format!("{host_ip}/24");
    let _ = run(&["sudo", "ip", "addr", "del", &cidr, "dev", interface]);
}

/// Best-effort: disable Wake-on-LAN so the NIC doesn't keep the PHY (and the
/// peer's carrier) alive after the interface is taken down. Per-NIC via ethtool
/// on Linux; macOS WoL is a system-wide pref (`pmset womp`), so there we rely on
/// admin-down alone (USB-Ethernet adapters drop link on `ifconfig down`).
fn disable_wol(interface: &str) {
    if macos() {
        return;
    }
    let _ = run(&["sudo", "ethtool", "-s", interface, "wol", "d"]);
}

/// Administratively bring the interface down — drops carrier on the peer.
fn admin_down(interface: &str) {
    if macos() {
        let _ = run(&["sudo", "ifconfig", interface, "down"]);
    } else {
        let _ = run(&["sudo", "ip", "link", "set", interface, "down"]);
    }
}

// ── modes ───────────────────────────────────────────────────────────────────

pub const MODES: [&str; 4] = ["netboot", "link", "ffx", "off"];

/// netboot mode: tear down ffx, start DHCP+TFTP (idempotent).
pub fn mode_netboot(
    target: &str,
    interface: &str,
    host_ip: &str,
    tftp_root: &str,
    opts: &crate::netboot::BootOptions,
) -> Result<()> {
    guard_primary_interface("mode netboot", interface)?;
    del_host_ll(interface);
    if crate::state::is_netboot_running(target) {
        return Ok(());
    }
    crate::netboot::start(target, interface, host_ip, tftp_root, opts)
}

/// link mode: assign just the IPv4 host IP to the interface — no DHCP/TFTP
/// daemon and no ffx link-local. This is the bare host side of the link, for
/// testing it up/down (`mode link` brings the host IP up, `mode off` releases
/// it). Idempotent. Note `mode off` only releases the IP; it does not force the
/// physical carrier down (the NIC can hold the link up for Wake-on-LAN — see
/// docs/netif.md).
pub fn mode_link(target: &str, interface: &str, host_ip: &str) -> Result<()> {
    guard_primary_interface("mode link", interface)?;
    if crate::state::is_netboot_running(target) {
        crate::netboot::stop(target)?;
    }
    del_host_ll(interface);
    configure_interface(interface, host_ip)
}

/// ffx mode: stop netboot first, then add the host IPv6 link-local.
pub fn mode_ffx(target: &str, interface: &str) -> Result<()> {
    guard_primary_interface("mode ffx", interface)?;
    if crate::state::is_netboot_running(target) {
        crate::netboot::stop(target)?;
    }
    add_host_ll(interface)
}

/// off: stop netboot, remove the ffx LL and any stale IPv4.
pub fn mode_off(target: &str, interface: &str, host_ip: &str) -> Result<()> {
    guard_primary_interface("mode off", interface)?;
    if crate::state::is_netboot_running(target) {
        crate::netboot::stop(target)?;
    }
    del_host_ll(interface);
    del_host_ip(interface, host_ip);
    Ok(())
}

/// Force the link down *hard*: do everything `mode off` does (stop netboot,
/// release the host IP + ffx LL), then disable Wake-on-LAN and admin-down the
/// interface so the peer sees carrier loss. `mode off` alone only releases the
/// host IP and can leave the carrier up (a WoL-capable NIC keeps the PHY
/// energized) — use this when you need the target to actually *detect* link
/// loss. Bring the link back with `mode link`/`mode netboot`; WoL stays disabled
/// until re-enabled (`ethtool -s <iface> wol g`) or the adapter is replugged.
pub fn down_hard(target: &str, interface: &str, host_ip: &str) -> Result<()> {
    guard_primary_interface("down-hard", interface)?;
    mode_off(target, interface, host_ip)?;
    disable_wol(interface);
    admin_down(interface);
    Ok(())
}

pub struct NetifStatus {
    pub mode: &'static str,
    pub carrier: bool,
    pub inet: Vec<String>,
    pub inet6: Vec<String>,
    pub peers: Vec<String>,
}

/// Probe the active mode (never stored): daemons running → netboot; host LL
/// present → ffx; the static host IP present (but no daemon/LL) → link; else
/// off. `carrier` is the physical link state (independent of the mode — it can
/// read up in `off` if the NIC keeps the PHY alive for Wake-on-LAN).
pub fn get_status(target: &str, interface: &str, host_ip: &str) -> NetifStatus {
    let netboot_running = crate::state::is_netboot_running(target);
    let (inet, inet6) = iface_addresses(interface);
    let has_ll = has_host_ll(&inet6);
    let has_host_ip = inet
        .iter()
        .any(|a| a.split('/').next().unwrap_or(a) == host_ip);
    let mode = if netboot_running {
        "netboot"
    } else if has_ll {
        "ffx"
    } else if has_host_ip {
        "link"
    } else {
        "off"
    };
    let peers = if mode == "ffx" {
        ipv6_peers(interface)
    } else {
        Vec::new()
    };
    NetifStatus {
        mode,
        carrier: is_interface_active(interface),
        inet,
        inet6,
        peers,
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn the_default_route_interface_is_primary() {
        assert!(is_default_route_interface("en0", Some("en0")));
        assert!(!is_default_route_interface("en14", Some("en0")));
        // No default route at all (an offline control host): nothing is primary.
        assert!(!is_default_route_interface("en0", None));
    }

    #[test]
    fn every_mode_change_refuses_the_primary_interface() {
        for what in [
            "mode netboot",
            "mode link",
            "mode ffx",
            "mode off",
            "down-hard",
        ] {
            let err = refuse_primary_interface(what, "en0", Some("en0"))
                .expect_err("the default-route interface must be refused");
            let msg = err.to_string();
            assert!(
                msg.starts_with(&format!("refusing netif {what} on 'en0'")),
                "{msg}"
            );
            assert!(msg.contains("system default route"), "{msg}");
        }
    }

    #[test]
    fn a_dedicated_adapter_passes_the_guard() {
        assert!(refuse_primary_interface("mode link", "en14", Some("en0")).is_ok());
        assert!(refuse_primary_interface("down-hard", "en14", None).is_ok());
    }

    #[test]
    fn ipv4_addrs_parses_the_one_line_ip_listing() {
        // Captured shape of `ip -4 -o addr show dev eth1`: `-o` joins the
        // lifetime line onto the address line with a backslash, and a second
        // (secondary) address is a second line.
        let sample = "\
3: eth1    inet 192.0.2.10/24 brd 192.0.2.255 scope global eth1\\       valid_lft forever preferred_lft forever
3: eth1    inet 198.51.100.7/16 brd 198.51.255.255 scope global secondary dynamic eth1\\       valid_lft 3242sec preferred_lft 3242sec
";
        assert_eq!(ipv4_addrs(sample), vec!["192.0.2.10/24", "198.51.100.7/16"]);
        assert!(ipv4_addrs("").is_empty());
        // Only `inet` entries count.
        assert!(ipv4_addrs("3: eth1    <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500\n").is_empty());
    }
}
